### PR TITLE
Add an onnx optimizer and NCHW input conversion support.

### PR DESCRIPTION
**Pull request by wenbingl**

### CODE REVIEW

File: tests/end2end/test_keras_converter.py [hunk 1]
```diff
@@ -19,8 +22,8 @@ def __init__(self, alpha=1.0, beta=1.0, **kwargs):
     def build(self, input_shape):
         super(ScaledTanh, self).build(input_shape)
 
-    def call(self, x):
-        return self.alpha * K.tanh(self.beta * x)
+    def call(self, inputs, **kwargs):
```
Comment (xadupre) [inline tests/end2end/test_keras_converter.py]: **kwargs: is it necessary?

File: tests/end2end/test_keras_converter.py [hunk 2]
```diff
@@ -44,15 +47,41 @@ def test_custom_op(self):
...
+        converted_model = onnxmltools.convert_keras(model, custom_conversion_functions={ScaledTanh: custom_activation})
+
         actual = model.predict(x)
         self.assertIsNotNone(actual)
 
-        converted_model = onnxmltools.convert_keras(model, custom_conversion_functions={ScaledTanh: custom_activation})
         self.assertIsNotNone(converted_model)
         # to check the model, you can print(str(converted_model))
-        dump_data_and_model(x.astype(np.float32), model, converted_model, basename="KerasCustomOp-Out0",
+        dump_data_and_model(x.astype(np.float32), model, converted_model, basename="Keras_custom_op-Out0",
```
Comment (xadupre) [inline tests/end2end/test_keras_converter.py]: Why rename it?

File: tests/end2end/test_keras_converter.py [hunk 3]
```diff
@@ -44,15 +47,41 @@ def test_custom_op(self):
...
+        model.add(MaxPooling2D((2, 2), strides=(2, 2), data_format='channels_last'))
+
+        model.compile(optimizer='sgd', loss='mse')
+        converted_model = onnxmltools.convert_keras(model, channel_first_inputs=[model.inputs[0].name])
+
+        expected = model.predict(x)
+        self.assertIsNotNone(expected)
+        self.assertIsNotNone(converted_model)
+
+        try:
```
Comment (xadupre) [inline tests/end2end/test_keras_converter.py]: Why not using dump_data_and_model?

File: tests/end2end/test_keras_converter.py (hunk 1 above)
Comment (wenbingl) [reply tests/end2end/test_keras_converter.py]: compliant to keras interface to kill PEP warning.

File: tests/end2end/test_keras_converter.py (hunk 3 above)
Comment (wenbingl) [reply tests/end2end/test_keras_converter.py]: The inputs between keras and onnx model are different, which hasn't been supported by dump_data_and_model.

Comment (xadupre) [reply tests/end2end/test_keras_converter.py]: Why not adding an extra parameter to support as transform (input_transform=lambda input: input.transpose( …)?

File: onnxmltools/convert/keras/convert.py [hunk 4]
```diff
@@ -31,19 +31,22 @@ def convert(model, name=None, initial_types=None, doc_string='', target_opset=No
     and a type defined in data_types.py.
     :param doc_string: A string attached onto the produced ONNX model
     :param target_opset: number, for example, 7 for ONNX 1.2, and 8 for ONNX 1.3.
+    :param channel_first_inputs, enforce the input format of the converted model is channel_first, i.e. NCHW
```
Comment (wschin) [inline onnxmltools/convert/keras/convert.py]: Would this be better?
```suggestion
    :param channel_first_inputs: specifies names of 4-D inputs which are forced to be in channel-first format (i.e., NCHW) in the converted model. It's a list of string; for example, ['input1', 'input2'].
```

**Comment by wschin** [comment]
My feeling is that you will end up with another `Topology` or `IR` for more sophisticated optimizations. Is it possible to reuse existing components?

**Pull request merged**